### ISSUE TITLE
Damage modeling decomposition dependent

### ISSUE THREAD

**Issue by theodordesousa** [peridigm/peridigm#119]
I am running a isotropic elastic-plastic hardening correspondence simulation with the interface aware model. 
The domain is the same as the tensile test example.

The damage seems to be divided into the same amount of regions as CPU cores I decompose the domain for. 
Could there be an error in my decomposition? I'm using the text_to_genesis script. 
Or could the damage model be defective?

![CG1_core8_dmg](https://user-images.githubusercontent.com/35833663/82211764-adf17e00-9911-11ea-9551-44e434076a6b.png)

![CG1_core20_dmg](https://user-images.githubusercontent.com/35833663/82211797-b649b900-9911-11ea-9e48-ccb145c5af42.png)

**Comment by johntfoster** [comment]
It's highly likely that the damage model is not communicating the damage
fields correctly across processor boundaries.

**Comment by theodordesousa** [comment]
Right, thank you. I will see if the same problem persists using the non-correspondence elastic plastic hardening model with critical stretch.

**Comment by theodordesousa** [comment]
The issue does not persist with the non-correspondence elastic plastic hardening model and regular critical stretch.

**Comment by andrisfreimanis** [comment]
I was writing a new damage model and noticed a similar behavior. I think that there is a problem accessing ELEMENT field data, when a neighbor does not belong 
* to the same process and
* to the same block. 

When interface aware damage model loops over neighbors of a node, it accesses neighbor's critical stretch parameter on line 152 here 
`const double neighborCriticalStretch = criticalStretch[neighborID];       
// Skip neighbors that do not allow bond breaking or bonds that are already broken
if(bondDamage[bondIndex]!=0.0 || neighborCriticalStretch == 0.0){bondIndex+=1;continue;}`
This works fine if both the node and its neighbor belongs to the same process and block. If not then criticalStretch[neighborID] returns 0. You can check it with a debugger, criticalStretch will be the value in the input file for all nodes belonging to the same process and the same block, but 0 for ghost nodes and nodes from another block. Then if statement on 154 checks the returned value, it's 0 and damage isn't even calculated.

[files.zip](https://github.com/peridigm/peridigm/files/4799430/files.zip)
You can see the problem by running the attached .yaml files on one and two cores and comparing the results.  "interfaceAwareOneBlockNP1.e" shows the correct damage, because all nodes and their neighbors belong to the same block and the same process. No damage can be seen in "interfaceAwareOneBlockNP2.e" and "interfaceAwareTwoBlocks.e", because of the bug.  

This probably has something to do with how data manager passes ELEMENT data between blocks and processes. Unfortunately this is roughly the limit of my Peridigm knowledge, so if someone could point me in the right direction I could investigate further.

**Comment by djlittl** [comment]
Correct, Peridigm does not synchronize all fields across process and block boundaries by default due to computational expense.  It only synchs a subset of the fields.  I'm pretty sure (80% confidence) that the material model function FieldIdsForSynchronizationAfterPrecompute() is the proper way to alter this behavior.  This function is the mechanism for a material model to inform the main driver routines in Peridigm that certain fields need to be sychronized.  By default it returns an empty list, but you should be able to add your field of interest to this list.  See the Material base class in Peridigm_Material.hpp for details.  I believe this will work for node data and element data, but will not work for bond data.  Keep in mind that if you have different material models assigned to different blocks, then the data fields (element data) allocated for those blocks could be different, so there's a potential for confusion there.

**Comment by djlittl** [comment]
Note on bond data:  I believe the only way to achieve the desired effect for bond data is to repeat the calculation on all relevant processors.  You need to ensure that the calculation will return the same result for a given bond on all processors.  This is how the critical stretch bond failure law works.
